streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java

```diff
@@ -359,15 +362,24 @@ public final InternalTopologyBuilder setApplicationId(final String applicationId
         return this;
     }
 
-    public synchronized final InternalTopologyBuilder setStreamsConfig(final StreamsConfig config) {
-        Objects.requireNonNull(config, "config can't be null");
-        this.config = config;
+    public synchronized final void setTopologyProperties(final Properties props) {
+        this.topologyProperties = props;
+    }
 
-        return this;
+    public synchronized final void setStreamsConfig(final StreamsConfig config) {
```

Comment:
This is a meta comment: I feel we can try to re-order the topology steps in a better way now. Today we have the following in order which looks awkward to me.

1) InternalStreamsBuilder#internalTopologyBuilder(props): this is to do the logical plan -> physical plan build. The `props` should be a per-topology overrides in the long run.
2) InternalStreamsBuilder#rewriteTopology(StreamsConfig config): this is to do some post-plan generation modifications which, ideally, should be part of 1) as well. Here the `config` is at global application-level.
3) Inside 2) we call `setStreamsConfig(config);` which is to set the global application configs.

I think a better order should be first setting the global configs, and then the per-topology overrides. So we should consider setting the global configs at the beginning when constructing the Topology (either from StreamsBuilder with DSL, or directly with PAPI), then the per-topology props for overrides.

Of course the reason we have this awkwardness is because for PAPI, the Topology is built without "props" passed in at all and hence we can only rely on the last step 3) to make sure the configs are finally initialized. But now with the new `NamedTopology` we can bypass this by requiring it to always come with a prop at the construction time as well, which would be used to set both (the func names are just for illustrations):

* internalTopologyBuilder.setTopologyOverrides(props);
* internalTopologyBuilder.setApplicationConfigs(config);

We would end up with the following scenarios:

1. No named-topology, DSL: props are set at StreamsBuilder.build(), and we know there's no overrides.
2. No named-topology, PAPI: the only pain left, as we have to wait at InternalTopologyBuilder.rewriteTopology, but still we know there's no overrides.
3. Named-topology, DSL: props are set at constructor, for both application-level and overrides.
4. Named-topology, PAPI: currently no one should be using in this way, but just in case it is supported it is still the same as 3) above.

And if in the long run we would stick with the NamedTopology where we would deprecate 1/2 above, we would end up naturally at the state where we enforce props to be set at the topology construction time.

WDYT?

Reply:
That all makes sense to me -- initially I was trying to make this just work with the existing public API, but ultimately figured that even the concept of "topology-level overrides" would require a KIP. If we're only dealing with named topologies then we're free to do things the right way from the start -- thanks for the suggestions :)

Reply:
@guozhangwang Played around with this a bit and was able to clean things up nicely, plus address some other awkwardness that was bugging me. Worked out great (just need to do some cleanup of the tests now...)